local_plugins/imports.py

```diff
@@ -0,0 +1,342 @@
+                self._error(node, ImportErrors.not_joined)
+                break
+
+    def check_wrapping(self, node: ast.ImportFrom) -> None:
+        start_tokens = self._filtered_tokens(node.lineno)
+        end_tokens = self._filtered_tokens(node.end_lineno)
+
+        if len(start_tokens) >= 4 and start_tokens[3].string == '(' and end_tokens[-1].string == ')':
+            if len(start_tokens) > 4 or len(end_tokens) > 1:
+                self._error(node, ImportErrors.missing_breaks)
+        else:
+            self._error(node, ImportErrors.not_wrapped)
+
+        for i in range(node.lineno + 1, node.end_lineno):
+            tokens = self._filtered_tokens(i)
+            if len(tokens) > 2:
+                self._error(node, ImportErrors.multiline)
+            elif tokens[-1].string != ',':
+                self._error(node, ImportErrors.no_trailing_comma)
+
+    def check_symbol_order(self, node: ast.ImportFrom):
+        for alias0, alias1 in zip(node.names, node.names[1:]):
+            if alias0.name > alias1.name:
+                self._error(node, ImportErrors.symbol_not_ordered)
+
+    def _error(self, node: EitherImport, err: ImportErrors) -> None:
+        self.errors.append(ErrorInfo(node.lineno, node.col_offset, err.value, None))
+
+    def _is_correct_order(self,
+                          node1: EitherImport,
+                          node2: EitherImport,
+                          order_info1: ModuleOrderInfo,
+                          order_info2: ModuleOrderInfo):
+        return (node1.lineno <= node2.lineno) == (order_info1 <= order_info2)
+
+    def _visited_successor(self, node: EitherImport) -> Optional[Tuple[EitherImport, ModuleOrderInfo]]:
+        """
+        Scan the the list of previously visisted nodes for the node with the
+        lowest line number that is greater than the provided node's line number.
+        """
+        return min(filter(lambda t: t[0].lineno > node.lineno, self.visited_order_info),
+                   key=lambda t: t[0].lineno,
+                   default=None)
+
+    def _visited_predecessor(self, node: EitherImport) -> Optional[Tuple[EitherImport, ModuleOrderInfo]]:
+        """
+        Scan the the list of previously visisted nodes for the node with the
+        highest line number that is less than the provided node's line number.
+        """
+        return max(filter(lambda t: t[0].lineno < node.lineno, self.visited_order_info),
+                   key=lambda t: t[0].lineno,
+                   default=None)
+
+    def _filtered_tokens(self, linenno):
+        return [tokeninfo
+                for tokeninfo in self.line_tokens[linenno - 1]  # 1-based indexing for source code lines
+                if tokeninfo.type not in (tokenize.COMMENT, tokenize.NEWLINE, tokenize.NL, tokenize.ENDMARKER)]
+
+    @classmethod
+    def demo(cls, source: str):
+        tokens = tokenize.generate_tokens((line + '\n' for line in source.split('\n')).__next__)
```

Comment:
I had to stare at this for a while to understand it. Maybe this is more readable:

```suggestion
        lines = (line + '\n' for line in source.split('\n'))

        def readline():
            return next(lines)

        tokens = tokenize.generate_tokens(readline)
```